doc/doxygen/man/man3/EmbData.3
.TH "EmbData" 3 "Thu Aug 17 2017" "Version 0.5.1" "steghide" \" -*- nroff -*-
.ad l
.nh
.SH NAME
EmbData
.SH SYNOPSIS
.br
.PP
.PP
\fC#include <EmbData\&.h>\fP
.SS "Public Types"

.in +1c
.ti -1c
.RI "enum \fBMODE\fP { \fBEMBED\fP, \fBEXTRACT\fP }"
.br
.ti -1c
.RI "enum \fBSTATE\fP { \fBREAD_MAGIC\fP, \fBREAD_VERSION\fP, \fBREAD_ENCINFO\fP, \fBREAD_NPLAINBITS\fP, \fBREAD_ENCRYPTED\fP, \fBEND\fP }"
.br
.in -1c
.SS "Public Member Functions"

.in +1c
.ti -1c
.RI "\fBEmbData\fP (\fBMODE\fP m, std::string pp, std::string fn='')"
.br
.ti -1c
.RI "\fBBitString\fP \fBgetBitString\fP (void)"
.br
.ti -1c
.RI "bool \fBfinished\fP (void)"
.br
.ti -1c
.RI "unsigned long \fBgetNumBitsRequested\fP (void)"
.br
.ti -1c
.RI "void \fBaddBits\fP (\fBBitString\fP addbits)"
.br
.ti -1c
.RI "void \fBsetEncAlgo\fP (\fBEncryptionAlgorithm\fP a)"
.br
.ti -1c
.RI "\fBEncryptionAlgorithm\fP \fBgetEncAlgo\fP (void) const"
.br
.ti -1c
.RI "void \fBsetEncMode\fP (\fBEncryptionMode\fP m)"
.br
.ti -1c
.RI "\fBEncryptionMode\fP \fBgetEncMode\fP (void) const"
.br
.ti -1c
.RI "void \fBsetCompression\fP (int c)"
.br
.ti -1c
.RI "int \fBgetCompression\fP (void) const"
.br
.ti -1c
.RI "void \fBsetChecksum\fP (bool c)"
.br
.ti -1c
.RI "bool \fBgetChecksum\fP (void) const"
.br
.ti -1c
.RI "bool \fBchecksumOK\fP (void) const"
.br
.ti -1c
.RI "void \fBsetData\fP (const std::vector< \fBBYTE\fP > data)"
.br
.ti -1c
.RI "std::vector< \fBBYTE\fP > \fBgetData\fP (void) const"
.br
.ti -1c
.RI "std::string \fBgetFileName\fP (void) const"
.br
.in -1c
.SS "Static Public Attributes"

.in +1c
.ti -1c
.RI "static const unsigned int \fBMinStegoHeaderSize\fP = 50"
.br
.RI "the minimum size of the part of the generatred \fBBitString\fP that is not the data "
.in -1c
.SS "Protected Member Functions"

.in +1c
.ti -1c
.RI "std::string \fBstripDir\fP (std::string s)"
.br
.in -1c
.SS "Private Attributes"

.in +1c
.ti -1c
.RI "\fBMODE\fP \fBMode\fP"
.br
.ti -1c
.RI "\fBSTATE\fP \fBState\fP"
.br
.ti -1c
.RI "unsigned long \fBNPlainBits\fP"
.br
.ti -1c
.RI "unsigned long \fBNumBitsRequested\fP"
.br
.RI "the number of bits that the caller must at least supply to addBits "
.ti -1c
.RI "unsigned long \fBNumBitsNeeded\fP"
.br
.RI "exactly the number of bits that the next step will consume from Reservoir and addBits together "
.ti -1c
.RI "\fBBitString\fP \fBReservoir\fP"
.br
.ti -1c
.RI "std::string \fBPassphrase\fP"
.br
.ti -1c
.RI "unsigned short \fBVersion\fP"
.br
.RI "version read from input bitstring "
.ti -1c
.RI "\fBEncryptionAlgorithm\fP \fBEncAlgo\fP"
.br
.ti -1c
.RI "\fBEncryptionMode\fP \fBEncMode\fP"
.br
.ti -1c
.RI "int \fBCompression\fP"
.br
.RI "compression level: 0(none),1(best speed),\&.\&.\&.,9(best compression) "
.ti -1c
.RI "bool \fBChecksum\fP"
.br
.RI "will a checksum be embedded ? "
.ti -1c
.RI "unsigned long \fBCRC32\fP"
.br
.RI "the checksum "
.ti -1c
.RI "std::string \fBFileName\fP"
.br
.ti -1c
.RI "std::vector< \fBBYTE\fP > \fBData\fP"
.br
.RI "contains the actual message to be embedded "
.in -1c
.SS "Static Private Attributes"

.in +1c
.ti -1c
.RI "static const unsigned int \fBNBitsNPlainBits\fP = 32"
.br
.RI "number of bits used to code the number of plain bits "
.ti -1c
.RI "static const unsigned int \fBNBitsNUncompressedBits\fP = 32"
.br
.RI "number of bits used to code the number of uncompressed bits "
.ti -1c
.RI "static const unsigned int \fBNBitsCrc32\fP = 32"
.br
.RI "size of a crc32 checksum in bits "
.ti -1c
.RI "static const unsigned short \fBCodeVersion\fP = 0"
.br
.RI "version of this steghide embedding (stego compatibility of \fBEmbData\fP) "
.ti -1c
.RI "static const \fBUWORD32\fP \fBMagic\fP = 0x73688DUL"
.br
.RI "steghide magic to recognize embedded data (the string 'shm') "
.ti -1c
.RI "static const unsigned int \fBNBitsMagic\fP = 24"
.br
.RI "size (in bits of Magic) "
.in -1c
.SH "Member Enumeration Documentation"
.PP 
.SS "enum \fBEmbData::MODE\fP"

.PP
\fBEnumerator\fP
.in +1c
.TP
\fB\fIEMBED \fP\fP
.TP
\fB\fIEXTRACT \fP\fP
.SS "enum \fBEmbData::STATE\fP"

.PP
\fBEnumerator\fP
.in +1c
.TP
\fB\fIREAD_MAGIC \fP\fP
.TP
\fB\fIREAD_VERSION \fP\fP
.TP
\fB\fIREAD_ENCINFO \fP\fP
.TP
\fB\fIREAD_NPLAINBITS \fP\fP
.TP
\fB\fIREAD_ENCRYPTED \fP\fP
.TP
\fB\fIEND \fP\fP
.SH "Constructor & Destructor Documentation"
.PP 
.SS "EmbData::EmbData (\fBMODE\fP m, std::string pp, std::string fn = \fC''\fP)"
construct a new \fBEmbData\fP object 
.PP
\fBParameters:\fP
.RS 4
\fIm\fP the mode (EMBED or EXTRACT) 
.br
\fIpp\fP the passphrase 
.br
\fIfn\fP the filename (only need for mode EMBED) 
.RE
.PP

.SH "Member Function Documentation"
.PP 
.SS "void EmbData::addBits (\fBBitString\fP addbits)"

.SS "bool EmbData::checksumOK (void) const"
check if crc32 checksum is ok (needs filled Data and CRC32 fields) 
.PP
\fBReturns:\fP
.RS 4
true iff checksum is ok 
.RE
.PP

.SS "bool EmbData::finished (void)"

.SS "\fBBitString\fP EmbData::getBitString (void)"

.SS "bool EmbData::getChecksum (void) const"

.SS "int EmbData::getCompression (void) const"

.SS "std::vector<\fBBYTE\fP> EmbData::getData (void) const\fC [inline]\fP"

.SS "\fBEncryptionAlgorithm\fP EmbData::getEncAlgo (void) const"

.SS "\fBEncryptionMode\fP EmbData::getEncMode (void) const"

.SS "std::string EmbData::getFileName (void) const\fC [inline]\fP"

.SS "unsigned long EmbData::getNumBitsRequested (void)"
get the minimum length of the \fBBitString\fP that is to be passed to addBits 
.SS "void EmbData::setChecksum (bool c)"

.SS "void EmbData::setCompression (int c)"

.SS "void EmbData::setData (const std::vector< \fBBYTE\fP > data)\fC [inline]\fP"

.SS "void EmbData::setEncAlgo (\fBEncryptionAlgorithm\fP a)"

.SS "void EmbData::setEncMode (\fBEncryptionMode\fP m)"

.SS "std::string EmbData::stripDir (std::string s)\fC [protected]\fP"

.SH "Member Data Documentation"
.PP 
.SS "bool EmbData::Checksum\fC [private]\fP"

.SS "const unsigned short EmbData::CodeVersion = 0\fC [static]\fP, \fC [private]\fP"

.SS "int EmbData::Compression\fC [private]\fP"

.SS "unsigned long EmbData::CRC32\fC [private]\fP"

.SS "std::vector<\fBBYTE\fP> EmbData::Data\fC [private]\fP"

.SS "\fBEncryptionAlgorithm\fP EmbData::EncAlgo\fC [private]\fP"

.SS "\fBEncryptionMode\fP EmbData::EncMode\fC [private]\fP"

.SS "std::string EmbData::FileName\fC [private]\fP"

.SS "const \fBUWORD32\fP EmbData::Magic = 0x73688DUL\fC [static]\fP, \fC [private]\fP"

.SS "const unsigned int EmbData::MinStegoHeaderSize = 50\fC [static]\fP"

.SS "\fBMODE\fP EmbData::Mode\fC [private]\fP"

.SS "const unsigned int EmbData::NBitsCrc32 = 32\fC [static]\fP, \fC [private]\fP"

.SS "const unsigned int EmbData::NBitsMagic = 24\fC [static]\fP, \fC [private]\fP"

.SS "const unsigned int EmbData::NBitsNPlainBits = 32\fC [static]\fP, \fC [private]\fP"

.SS "const unsigned int EmbData::NBitsNUncompressedBits = 32\fC [static]\fP, \fC [private]\fP"

.SS "unsigned long EmbData::NPlainBits\fC [private]\fP"

.SS "unsigned long EmbData::NumBitsNeeded\fC [private]\fP"

.SS "unsigned long EmbData::NumBitsRequested\fC [private]\fP"

.SS "std::string EmbData::Passphrase\fC [private]\fP"

.SS "\fBBitString\fP EmbData::Reservoir\fC [private]\fP"

.SS "\fBSTATE\fP EmbData::State\fC [private]\fP"

.SS "unsigned short EmbData::Version\fC [private]\fP"


.SH "Author"
.PP 
Generated automatically by Doxygen for steghide from the source code\&.


src/EmbData.cc
/*
 * steghide 0.5.1 - a steganography program
 * Copyright (C) 1999-2003 Stefan Hetzl <shetzl@chello.at>
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; either version 2
 * of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.
 *
 */

#include "AUtils.h"
#include "BinaryIO.h"
#include "BitString.h"
#include "EmbData.h"
#include "error.h"
#include "MCryptPP.h"
#include "MHashPP.h"
#include "common.h"

EmbData::EmbData (MODE m, std::string pp, std::string fn)
	: Mode(m), Passphrase(pp), FileName(fn)
{
	//std::cout << "embdata!!!!!" << std::endl;
	if (m == EXTRACT) {
		NumBitsNeeded = NumBitsRequested = NBitsMagic ;
		Version = CodeVersion ;
		State = READ_MAGIC ;
		Reservoir = BitString() ;
	}
}

bool EmbData::finished ()
{
	myassert (Mode == EXTRACT) ;
	return (State == END) ;
}

unsigned long EmbData::getNumBitsRequested ()
{
	myassert (Mode == EXTRACT) ;
	return NumBitsRequested ;
}

void EmbData::addBits (BitString addbits)
{
	myassert (Mode == EXTRACT) ;

#ifdef DEBUG
	printDebug (1, "\nEmbData::addBits called with") ;
	printDebug (1, " addbits:") ;
	addbits.printDebug (1, 2) ;
	printDebug (1, " Reservoir:") ;
	Reservoir.printDebug (1, 2) ;
#endif

	Reservoir.append(addbits);
	//addbits.print();
	BitString bits ;
	if (Reservoir.getLength() >= NumBitsNeeded) {
		bits = Reservoir.cutBits (0, NumBitsNeeded) ; // take exactly the first NumBitsNeeded bits from Reservoir | addbits
	}
	else { // Reservoir.getLength() < NumBitsNeeded
		myassert(false) ;
	}



#ifdef DEBUG
	printDebug (1, "bits is now:") ;
	bits.printDebug (1, 2) ;
#endif
	//std::cout << "switch" << std::endl;
	switch (State) {
		case READ_MAGIC:
		{
			//std::cout << "magic" << std::endl;
#ifdef DEBUG
			printDebug (1, "in the READ_MAGIC state") ;
#endif
			if (bits.getValue(0, NBitsMagic) == Magic) {
				std::cout << "read magic ************" << std::endl;
				NumBitsNeeded = 1 ;
				NumBitsRequested = AUtils::bminus<unsigned long> (NumBitsNeeded, Reservoir.getLength()) ;
				State = READ_VERSION ;
			}
			else {
				//std::cout << "magic error" << std::endl;
				throw SteghideError (_("could not extract any data with that passphrase!")) ;
			}
			break ;
		}	

		case READ_VERSION:
		{
			std::cout << "read version" << std::endl;
#ifdef DEBUG
			printDebug (1, "in the READ_VERSION state") ;
#endif
			if (bits[0] == true) {
				Version++ ;
				NumBitsNeeded = AUtils::bminus ((UWORD32) 1, Reservoir.getLength()) ;
			}
			else {
				if (Version > CodeVersion) {
					throw CorruptDataError (_("attempting to read an embedding of version %d but steghide %s only supports embeddings of version %d."), Version, VERSION, CodeVersion) ;
				}
				NumBitsNeeded = EncryptionAlgorithm::IRep_size + EncryptionMode::IRep_size ;
				NumBitsRequested = AUtils::bminus<unsigned long> (NumBitsNeeded, Reservoir.getLength()) ;
				State = READ_ENCINFO ;
			}
			break ;
		}

		case READ_ENCINFO: {
			std::cout << "read encinfo" << std::endl;
#ifdef DEBUG
			printDebug (1, "in the READ_ENCINFO state") ;
#endif

			unsigned int algo = (unsigned int) bits.getValue (0, EncryptionAlgorithm::IRep_size) ;
			if (EncryptionAlgorithm::isValidIntegerRep (algo)) {
				EncAlgo.setValue ((EncryptionAlgorithm::IRep) algo) ;
			}
			unsigned int mode = (unsigned int) bits.getValue (EncryptionAlgorithm::IRep_size, EncryptionMode::IRep_size) ;
			if (EncryptionMode::isValidIntegerRep (mode)) {
				EncMode.setValue ((EncryptionMode::IRep) mode) ;
			}

			NumBitsNeeded = NBitsNPlainBits ;
			NumBitsRequested = AUtils::bminus<unsigned long> (NumBitsNeeded, Reservoir.getLength()) ;
			State = READ_NPLAINBITS ;

#ifndef USE_LIBMCRYPT
			if (EncAlgo.getIntegerRep() != EncryptionAlgorithm::NONE) {
				throw SteghideError (_(
							"The embedded data is encrypted but steghide has been compiled without encryption\n"
							"support. To be able to read the embedded data, you have to install libmcrypt\n"
							"(http://mcrypt.sourceforge.net/) and recompile steghide.")) ;
			}
#endif
		break ; }

		case READ_NPLAINBITS: {
			std::cout << "read plainbits" << std::endl;
#ifdef DEBUG
			printDebug (1, "in the READ_NPLAINBITS state") ;
#endif
			
			NPlainBits = bits.getValue (0, NBitsNPlainBits) ;

#ifdef USE_LIBMCRYPT
			NumBitsNeeded = (UWORD32) MCryptPP::getEncryptedSize (EncAlgo, EncMode, NPlainBits) ;
#else
			NumBitsNeeded = NPlainBits ;
#endif
			NumBitsRequested = AUtils::bminus<unsigned long> (NumBitsNeeded, Reservoir.getLength()) ;

			State = READ_ENCRYPTED ;
		break ; }

		case READ_ENCRYPTED: {
			std::cout << "read encrypted" << std::endl;
#ifdef DEBUG
			printDebug (1, "in the READ_ENCRYPTED state") ;
#endif

			BitString plain ;
#ifdef USE_LIBMCRYPT
			if (EncAlgo.getIntegerRep() == EncryptionAlgorithm::NONE) {
				plain = bits ;
			}
			else {
				MCryptPP crypto (EncAlgo, EncMode) ;
				plain = crypto.decrypt (bits, Passphrase) ;
			}
#else
			plain = bits ;
#endif

			plain.truncate (0, NPlainBits) ;	// cut off random padding used to achieve full number of encryption blocks

			unsigned long pos = 0 ;

			// read Compression (and uncompress)
			Compression = ((plain[pos++]) ? 9 : 0) ; // to make compression contain a value that makes sense
#ifdef DEBUG
			printDebug (2, " compression: %d\n", plain[pos - 1]) ;
#endif
			if (Compression > 0) {
				UWORD32 NUncompressedBits = plain.getValue (pos, NBitsNUncompressedBits) ;
#ifdef DEBUG
				printDebug (2, " nuncobits: %lu\n", NUncompressedBits) ;
#endif
				pos += NBitsNUncompressedBits ;

				plain.truncate (pos, plain.getLength()) ;
				pos = 0 ;
				plain.uncompress (NUncompressedBits) ;
			}

			std::cout << "plain text" << std::endl;
			plain.print();

			// read Checksum
			Checksum = plain[pos++] ;
#ifdef DEBUG
			printDebug (2, " checksum: %d\n", plain[pos - 1]) ;
#endif
			if (Checksum) {
				CRC32 = plain.getValue (pos, NBitsCrc32) ;
#ifdef DEBUG
				printDebug (2, " crc32: 0x%x\n", CRC32) ;
#endif
				pos += NBitsCrc32 ;
			}

			// read filename
			char curchar = '\0' ;
			FileName = "" ;
			do {
				curchar = (char) plain.getValue (pos, 8) ;
				if (curchar != '\0') {
					FileName += curchar ;
				}
				pos += 8 ;
			} while (curchar != '\0') ;

			// extract data
			if ((plain.getLength() - pos) % 8 != 0) {
				throw CorruptDataError (_("the embedded data has an invalid length.")) ;
			}
			const unsigned long extdatalen = (plain.getLength() - pos) / 8 ;
			Data.resize (extdatalen) ;
			for (unsigned int i = 0 ; i < extdatalen ; i++) {
				Data[i] = ((BYTE) plain.getValue (pos, 8)) ;
				pos += 8 ;
			}

			NumBitsNeeded = 0 ;
			NumBitsRequested = 0 ;
			State = END ;
		break ; }

		case END:
		default: {
			myassert (0) ;
		break ; }
	}
}

bool EmbData::checksumOK (void) const
{
	// test if checksum is ok
	bool ok = true ;
	if (Checksum) {
		MHashPP hash (MHASH_CRC32) ;
		for (std::vector<BYTE>::const_iterator i = Data.begin() ; i != Data.end() ; i++) {
			hash << *i ;
		}
		hash << MHashPP::endhash ;
		unsigned long calccrc32 = hash.getHashBits().getValue(0, NBitsCrc32) ;

		if (calccrc32 == CRC32) {
			ok = true ;
		}
		else {
			ok = false ;
		}
	}
	return ok ;
}

void EmbData::setEncAlgo (EncryptionAlgorithm a)
{
	EncAlgo = a ;
}
	
EncryptionAlgorithm EmbData::getEncAlgo () const
{
	return EncAlgo ;
}
	
void EmbData::setEncMode (EncryptionMode m) 
{
	EncMode = m ;
}
	
EncryptionMode EmbData::getEncMode () const
{
	return EncMode ;
}
	
void EmbData::setCompression (int c)
{
	Compression = c ;
}
	
int EmbData::getCompression (void) const
{
	return Compression ;
}
	
void EmbData::setChecksum (bool c)
{
	Checksum = c ;
}
	
bool EmbData::getChecksum (void) const
{
	return Checksum ;
}

BitString EmbData::getBitString ()
{
	myassert (Mode == EMBED) ;

	// assembling data that can be compressed
	BitString compr ;

	compr.append (Checksum) ;
	std::cout << "checksum" <<std::endl;
	compr.print();

	if (Checksum) {
		MHashPP hash (MHASH_CRC32) ;
		std::cout << "hash" <<std::endl;
		for (std::vector<BYTE>::iterator i = Data.begin() ; i != Data.end() ; i++) {

			hash << *i ;

		}
		hash << MHashPP::endhash ;
		hash.getHashBits().print();
		compr.append (hash.getHashBits()) ;
		std::cout << "hash" <<std::endl;
		compr.print();
	}
	
	compr.append (stripDir (FileName)) ;
	compr.append ((BYTE) 0, 8) ; // end of fileame
	
	compr.append (Data) ;
	std::string s = std::string(Data.begin(), Data.end());
	std::cout << s << std::endl;


	std::cout << "compression" <<std::endl;
	compr.print();
	// assembling data that can be encrypted
	BitString plain ;
	plain.append ((Compression > 0) ? true : false) ;
	plain.print();

	if (Compression > 0) {
		std::cout << "compression!" <<std::endl;
		plain.append (compr.getLength(), NBitsNUncompressedBits) ;
		std::cout << compr.getLength() << std::endl;
		std::cout << Compression << std::endl;
		compr.compress (Compression) ;
	}
	compr.print();
	plain.print();
	plain.append (compr) ;
	plain.print();

	// put it all together
	BitString main ;
	std::cout << Magic <<std::endl;
	main.append(Magic, NBitsMagic) ;
	main.print();
	for (unsigned short i = 0 ; i < CodeVersion ; i++) {
		main.append(true) ;
	}
	main.append(false) ; // end of version
	main.print();
	main.append((UWORD16) EncAlgo.getIntegerRep(), EncryptionAlgorithm::IRep_size) ;
	main.append((UWORD16) EncMode.getIntegerRep(), EncryptionMode::IRep_size) ;
	main.append(plain.getLength(), NBitsNPlainBits) ;

#ifdef USE_LIBMCRYPT
	if (EncAlgo.getIntegerRep() != EncryptionAlgorithm::NONE) {
		MCryptPP crypto (EncAlgo, EncMode) ;
		plain = crypto.encrypt (plain, Passphrase) ;
	}
#else
	myassert (EncAlgo.getIntegerRep() == EncryptionAlgorithm::NONE) ;
#endif

	main.append (plain) ;

	return main ;
}

std::string EmbData::stripDir (std::string s)
{
	unsigned int start = 0 ;
	if ((start = s.find_last_of ("/\\")) == std::string::npos) {
		start = 0 ;
	}
	else {
		start += 1 ;
	}
	return s.substr (start, std::string::npos) ;
}
